### PR TITLE
Adding system prune for podman v2

### PR DESCRIPTION
**Pull request by sujil02**
Register system prune routes, handler to support system prune
Adds testcase to validate the system prune flow.

Signed-off-by: Sujil02 <sushah@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM

File: go.sum [hunk 1]
```diff
@@ -624,6 +624,7 @@ k8s.io/klog v1.0.0 h1:Pt+yjF5aB1xDSVbau4VsWe+dQNzA0qv1LlXdC2dF6Q8=
 k8s.io/klog v1.0.0/go.mod h1:4Bi6QPql/J/LkTDqv7R/cd3hPo4k2DG6Ptcz060Ez5I=
 k8s.io/kube-openapi v0.0.0-20190228160746-b3a7cee44a30/go.mod h1:BXM9ceUBTj2QnfH2MK1odQs778ajze1RxcmP6S8RVVc=
 k8s.io/kube-openapi v0.0.0-20191107075043-30be4d16710a/go.mod h1:1TqjTSzOxsLGIKfj0lK8EeCP7K1iUG65v09OM0/WG5E=
+k8s.io/kubernetes v1.13.0 h1:qTfB+u5M92k2fCCCVP2iuhgwwSOv1EkAkvQY1tQODD8=
```
Comment (TomSweeneyRedHat) [inline go.sum:627]: Just curious as to why/how this was added?

Comment (rhatdan) [reply go.sum:627]: This came in via make vendor.  Not sure why.

File: pkg/api/handlers/libpod/system.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	}{}
+
+	if err := decoder.Decode(&query, r.URL.Query()); err != nil {
+		utils.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest,
+			errors.Wrapf(err, "Failed to parse parameters for %s", r.URL.String()))
+		return
+	}
+
+	var systemPruneReport = new(entities.SystemPruneReport)
+	fmt.Println("in sytem prune")
```
Comment (jwhonce) [inline pkg/api/handlers/libpod/system.go:32]:
```suggestion
```

File: pkg/api/handlers/libpod/system.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		return
+	}
+
+	report := entities.ImagePruneReport{
+		Report: entities.Report{
+			Id:  results,
+			Err: nil,
+		},
+		Size: 0,
+	}
```
Comment (jwhonce) [inline pkg/api/handlers/libpod/system.go:60]:
```suggestion
	report := entities.ImagePruneReport{
		Report: entities.Report{
			Id:  results,
		},
	}
```
Don't initialize zero-value fields.

File: pkg/api/server/register_system.go [hunk 4]
```diff
@@ -4,12 +4,28 @@ import (
...
 	"github.com/containers/libpod/pkg/api/handlers/compat"
+	"github.com/containers/libpod/pkg/api/handlers/libpod"
 	"github.com/gorilla/mux"
 )
 
 func (s *APIServer) registerSystemHandlers(r *mux.Router) error {
 	r.Handle(VersionedPath("/system/df"), s.APIHandler(compat.GetDiskUsage)).Methods(http.MethodGet)
 	// Added non version path to URI to support docker non versioned paths
 	r.Handle("/system/df", s.APIHandler(compat.GetDiskUsage)).Methods(http.MethodGet)
+	// ................. .... ....................
```
Comment (jwhonce) [inline pkg/api/server/register_system.go:15]:
```suggestion
	// swagger:operation POST /libpod/system/prune libpod pruneSystem
	// ---
	// tags:
	//  - system
```

File: pkg/api/handlers/libpod/system.go (hunk 2 above)
Comment (sujil02) [reply pkg/api/handlers/libpod/system.go:32]: Fixed.

File: pkg/api/server/register_system.go (hunk 4 above)
Comment (sujil02) [reply pkg/api/server/register_system.go:15]: fixed.

File: pkg/api/handlers/libpod/system.go (hunk 3 above)
Comment (sujil02) [reply pkg/api/handlers/libpod/system.go:60]: Fixed here. @jwhonce should I fix it here as well? https://github.com/containers/libpod/blob/master/pkg/domain/infra/abi/images.go#L49

Comment (jwhonce) [reply pkg/api/handlers/libpod/system.go]: @sujil02 Please do.

Comment (sujil02) [reply pkg/api/handlers/libpod/system.go]: @jwhonce done. Please re-review so I can get the ABI and tunnel flow in.

File: pkg/api/server/register_system.go [hunk 5]
```diff
@@ -4,12 +4,28 @@ import (
...
+	"github.com/containers/libpod/pkg/api/handlers/libpod"
 	"github.com/gorilla/mux"
 )
 
 func (s *APIServer) registerSystemHandlers(r *mux.Router) error {
 	r.Handle(VersionedPath("/system/df"), s.APIHandler(compat.GetDiskUsage)).Methods(http.MethodGet)
 	// Added non version path to URI to support docker non versioned paths
 	r.Handle("/system/df", s.APIHandler(compat.GetDiskUsage)).Methods(http.MethodGet)
+	// Swagger:operation POST /libpod/system/prune libpod pruneSystem
+	// summary: Prune unused data
```
Comment (jwhonce) [inline pkg/api/server/register_system.go:16]:
```suggestion
        // ---
	// tags:
	//   - system
	// summary: Prune unused data
```

File: pkg/api/server/register_system.go [hunk 6]
```diff
@@ -4,12 +4,28 @@ import (
...
+	// responses:
+	//   200:
+	//     $ref: '#/responses/SystemPruneReport'
+	//   400:
+	//     $ref: "#/responses/BadParamError"
+	//   500:
+	//     $ref: "#/responses/InternalError"
+	r.Handle(VersionedPath("/libpod/system/prune"), s.APIHandler(libpod.SystemPrune)).Methods(http.MethodPost)
+	// Added non version path to URI to support docker non versioned paths
+	r.Handle("/libpod/system/prune", s.APIHandler(libpod.SystemPrune)).Methods(http.MethodPost)
```
Comment (jwhonce) [inline pkg/api/server/register_system.go:28]:
```suggestion
```
We only do unversioned for compatibility end points

File: pkg/bindings/system/system.go [hunk 7]
```diff
@@ -59,3 +61,26 @@ func Events(ctx context.Context, eventChan chan (handlers.Event), cancelChan cha
...
+	)
+	conn, err := bindings.GetClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	params := url.Values{}
+	if all != nil {
+		params.Set("All", strconv.FormatBool(*all))
+	}
+	if all != nil {
```
Comment (jwhonce) [inline pkg/bindings/system/system.go:78]:
```suggestion
	if volumes != nil {
```

File: pkg/bindings/test/system_test.go [hunk 8]
```diff
@@ -48,4 +56,90 @@ var _ = Describe("Podman system", func() {
...
+		Expect(len(systemPruneResponse.ContainerPruneReport.ID)).To(Equal(1))
+		Expect(len(systemPruneResponse.ImagePruneReport.Report.Id)).
+			To(BeNumerically(">", 0))
+		// Alpine image should not be pruned as used by running container
+		Expect(systemPruneResponse.ImagePruneReport.Report.Id).
+			ToNot(ContainElement("docker.io/library/alpine:latest"))
+		Expect(len(systemPruneResponse.VolumePruneReport)).To(Equal(0))
+	})
+
+	It("podman system prune running alpine container volume prune", func() {
```
Comment (jwhonce) [inline pkg/bindings/test/system_test.go:113]: This test didn't find coding error at https://github.com/containers/libpod/pull/5907/files#diff-d20431f0bfae3cfadda9b39a8ab6b074R78

File: pkg/api/server/register_system.go (hunk 6 above)
Comment (sujil02) [reply pkg/api/server/register_system.go]: Done.

File: pkg/bindings/test/system_test.go (hunk 8 above)
Comment (sujil02) [reply pkg/bindings/test/system_test.go:118]: Thank you for the catch there. Fixed the test to validate the use case https://github.com/containers/libpod/pull/5907/commits/0820e7c82abd1b84b3e81627f4ed396fb0f339bd#diff-745bc12747465f015915254b3f4917e1R101-R115

File: pkg/bindings/system/system.go (hunk 7 above)
Comment (sujil02) [reply pkg/bindings/system/system.go]: Fixed.

File: pkg/api/server/register_system.go (hunk 5 above)
Comment (sujil02) [reply pkg/api/server/register_system.go:19]: fixed.

**Pull request merged**